index out of range panic when setting tidb_opt_range_max_size small

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```
use test;
CREATE TABLE `sbtest1` (
	`id` bigint(20) NOT NULL AUTO_INCREMENT,
	`k` int(11) NOT NULL DEFAULT '0',
	`c` char(120) NOT NULL DEFAULT '',
	`pad` char(60) NOT NULL DEFAULT '',
	PRIMARY KEY (`id`) /*T![clustered_index] CLUSTERED */ ,
	KEY `k_1` (`k`)
);
set @@tidb_opt_range_max_size = 111;
explain select * from test.sbtest1 a where pad in ('1','1','1','1','1') and id in (1,1,1,1,1);
```

### 2. What did you expect to see? (Required)
Output the plan successfully. No panic happens.

### 3. What did you see instead (Required)
```
mysql> explain select * from test.sbtest1 a where pad in ('1','1','1','1','1') and id in (1,1,1,1,1);
ERROR 1105 (HY000): runtime error: index out of range [0] with length 0
```

error stack:
```
[2023/04/18 10:38:08.990 +08:00] [ERROR] [conn.go:1039] ["connection running loop panic"] [conn=2199023255955] [lastSQL="explain select * from test.sbtest1 a where pad in ('1','1','1','1','1') and id in (1,1,1,1,1)"] [err="runtime error: index out of range [0] with length 0"] [stack="github.com/pingcap/tidb/server.(*clientConn).Run.func1
	/Users/xuyifan/tidb/server/conn.go:1042
runtime.gopanic
	/usr/local/go/src/runtime/panic.go:884
github.com/pingcap/tidb/executor.(*Compiler).Compile.func1
	/Users/xuyifan/tidb/executor/compiler.go:54
runtime.gopanic
	/usr/local/go/src/runtime/panic.go:884
runtime.goPanicIndex
	/usr/local/go/src/runtime/panic.go:113
github.com/pingcap/tidb/util/ranger.(*Range).PrefixEqualLen
	/Users/xuyifan/tidb/util/ranger/types.go:220
github.com/pingcap/tidb/statistics.getPseudoRowCountByIndexRanges
	/Users/xuyifan/tidb/statistics/table.go:1198
github.com/pingcap/tidb/statistics.(*HistColl).GetRowCountByIndexRanges
	/Users/xuyifan/tidb/statistics/table.go:629
github.com/pingcap/tidb/planner/core.(*DataSource).detachCondAndBuildRangeForPath
	/Users/xuyifan/tidb/planner/core/logical_plans.go:1398
github.com/pingcap/tidb/planner/core.(*DataSource).fillIndexPath
	/Users/xuyifan/tidb/planner/core/logical_plans.go:1545
github.com/pingcap/tidb/planner/core.(*DataSource).DeriveStats
	/Users/xuyifan/tidb/planner/core/stats.go:422
github.com/pingcap/tidb/planner/core.(*baseLogicalPlan).recursiveDeriveStats
	/Users/xuyifan/tidb/planner/core/stats.go:130
github.com/pingcap/tidb/planner/core.(*baseLogicalPlan).recursiveDeriveStats
	/Users/xuyifan/tidb/planner/core/stats.go:123
github.com/pingcap/tidb/planner/core.physicalOptimize
	/Users/xuyifan/tidb/planner/core/optimizer.go:1112
github.com/pingcap/tidb/planner/core.DoOptimize
	/Users/xuyifan/tidb/planner/core/optimizer.go:305
github.com/pingcap/tidb/planner.optimize
	/Users/xuyifan/tidb/planner/optimize.go:469
github.com/pingcap/tidb/planner.Optimize
	/Users/xuyifan/tidb/planner/optimize.go:297
github.com/pingcap/tidb/planner/core.(*PlanBuilder).buildExplain
	/Users/xuyifan/tidb/planner/core/planbuilder.go:4994
github.com/pingcap/tidb/planner/core.(*PlanBuilder).Build
	/Users/xuyifan/tidb/planner/core/planbuilder.go:794
github.com/pingcap/tidb/planner.buildLogicalPlan
	/Users/xuyifan/tidb/planner/optimize.go:513
github.com/pingcap/tidb/planner.optimize
	/Users/xuyifan/tidb/planner/optimize.go:434
github.com/pingcap/tidb/planner.Optimize
	/Users/xuyifan/tidb/planner/optimize.go:297
github.com/pingcap/tidb/executor.(*Compiler).Compile
	/Users/xuyifan/tidb/executor/compiler.go:98
github.com/pingcap/tidb/session.(*session).ExecuteStmt
	/Users/xuyifan/tidb/session/session.go:2154
github.com/pingcap/tidb/server.(*TiDBContext).ExecuteStmt
	/Users/xuyifan/tidb/server/driver_tidb.go:252
github.com/pingcap/tidb/server.(*clientConn).handleStmt
	/Users/xuyifan/tidb/server/conn.go:2061
github.com/pingcap/tidb/server.(*clientConn).handleQuery
	/Users/xuyifan/tidb/server/conn.go:1852
github.com/pingcap/tidb/server.(*clientConn).dispatch
	/Users/xuyifan/tidb/server/conn.go:1339
github.com/pingcap/tidb/server.(*clientConn).Run
	/Users/xuyifan/tidb/server/conn.go:1120
github.com/pingcap/tidb/server.(*Server).onConn
	/Users/xuyifan/tidb/server/server.go:677"]
```

### 4. What is your TiDB version? (Required)
master

--- a/planner/core/integration_test.go
+++ b/planner/core/integration_test.go
@@ -5275,3 +5275,16 @@ func TestIssue41458(t *testing.T) {
 		require.Equalf(t, expectedRes[i], op, fmt.Sprintf("Mismatch at index %d.", i))
 	}
 }
+
+func TestIssue43116(t *testing.T) {
+	store := testkit.CreateMockStore(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+	tk.MustExec("CREATE TABLE `sbtest1` ( `id` bigint(20) NOT NULL AUTO_INCREMENT, `k` int(11) NOT NULL DEFAULT '0', `c` char(120) NOT NULL DEFAULT '', `pad` char(60) NOT NULL DEFAULT '', PRIMARY KEY (`id`) /*T![clustered_index] CLUSTERED */ , KEY `k_1` (`k`) )")
+	tk.MustExec("set @@tidb_opt_range_max_size = 111")
+	tk.MustQuery("explain format='brief' select * from test.sbtest1 a where pad in ('1','1','1','1','1') and id in (1,1,1,1,1)").Check(testkit.Rows(
+		"TableReader 8000.00 root  data:Selection",
+		"└─Selection 8000.00 cop[tikv]  in(test.sbtest1.id, 1, 1, 1, 1, 1), in(test.sbtest1.pad, \"1\", \"1\", \"1\", \"1\", \"1\")",
+		"  └─TableFullScan 10000.00 cop[tikv] table:a keep order:false, stats:pseudo"))
+	tk.MustQuery("show warnings").Check(testkit.Rows("Warning 1105 Memory capacity of 111 bytes for 'tidb_opt_range_max_size' exceeded when building ranges. Less accurate ranges such as full range are chosen"))
+}

--- a/util/ranger/points.go
+++ b/util/ranger/points.go
@@ -171,7 +171,7 @@ func FullRange() Ranges {
 
 // FullNotNullRange is (-∞, +∞) for Range.
 func FullNotNullRange() Ranges {
-	return Ranges{{LowVal: []types.Datum{types.MinNotNullDatum()}, HighVal: []types.Datum{types.MaxValueDatum()}}}
+	return Ranges{{LowVal: []types.Datum{types.MinNotNullDatum()}, HighVal: []types.Datum{types.MaxValueDatum()}, Collators: collate.GetBinaryCollatorSlice(1)}}
 }
 
 // NullRange is [null, null] for Range.

